### PR TITLE
HH should not process dropped nodes

### PR DESCRIPTION
**Pull request by otoolep**

### CODE REVIEW

**Comment by otoolep** [comment]
@jwilder @corylanou 

As for testing, I thought the simplest thing to do was to add the new code to the existing test. Yes, it means a somewhat longer test case, but it's very straightforward. This is the limit to which I push iot though.

**Comment by otoolep** [comment]
Patch updated to remove the stubbed-out "purge" call. All data deletion is left to simple expiration so that data is never dropped directly due to operations such as DROP SERVER.

**Comment by otoolep** [comment]
I see that with this change, the system will continually adding a new segment for all queues it purges. That is not what I wanted. Let me fix this up.

File: services/hh/processor.go [hunk 1]
```diff
@@ -273,6 +281,20 @@ func (p *Processor) updateShardStats(shardID uint64, stat string, inc int64) {
 	m.Add(stat, inc)
 }
 
+func (p *Processor) activeNodes() ([]uint64, error) {
```
Comment (otoolep) [inline services/hh/processor.go]: This could also be `activeQueues`, I could go either way.

**Comment by otoolep** [comment]
OK, this is ready to go @jwilder and @corylanou 

Lots of way to solve this, I think this should do it nicely.

File: services/hh/processor.go [hunk 2]
```diff
@@ -284,3 +306,41 @@ func (p *Processor) PurgeOlderThan(when time.Duration) error {
...
+		last, err := queue.LastModified()
+		if err != nil {
+			return err
+		}
+		if last.Before(time.Now().Add(-when)) {
+			// Close and remove the queue.
+			if err := queue.Close(); err != nil {
+				return err
+			}
+			if err := os.RemoveAll(queue.dir); err != nil {
```
Comment (jwilder) [inline services/hh/processor.go]: Seems like this be a func on `queue`.  `queue.Remove()`

File: services/hh/queue.go [hunk 3]
```diff
@@ -181,6 +181,17 @@ func (l *queue) PurgeOlderThan(when time.Time) error {
 	}
 }
 
+// LastModified returns the last time the queue was modified.
+func (l *queue) LastModified() (time.Time, error) {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	if len(l.segments) > 0 {
+		return l.segments[len(l.segments)-1].lastModified()
```
Comment (jwilder) [inline services/hh/queue.go]: Should use `l.tail` instead of working w/ `l.segments` directly.

Comment (jwilder) [inline services/hh/queue.go]: `l.tail` could be `nil` too if the queue is not open.

File: services/hh/service.go [hunk 4]
```diff
@@ -165,8 +172,19 @@ func (s *Service) expireWrites() {
...
+func (s *Service) deleteQueues() {
+	defer s.wg.Done()
+	ticker := time.NewTicker(time.Hour)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.closing:
+			return
+		case <-ticker.C:
+			if err := s.HintedHandoff.DeleteOlderThan(time.Duration(s.cfg.MaxAge)); err != nil {
```
Comment (jwilder) [inline services/hh/service.go]: `PurgeOlderThan` and `DeleteOlderThan` are very similarly named and might cause confusion in the future.  Maybe `PurgeInactiveOlderThan`?  Dunno...

**Comment by jwilder** [comment]
A could of small things but otherwise LGTM. :+1:

File: services/hh/processor.go (hunk 2 above)
Comment (otoolep) [inline services/hh/processor.go]: Yeah, I thought about that, but wasn't sure. Should `Remove()` work on an open-queue? Should the queue be closed first? What are the semantic of this command?

All these reasons made me fall back on having this operation occur outside of the queue. (I started with having it as a method on the queue, as you suggest). I don't feel super strongly about this, and wasn't too happy with the code acting on `path`, technically an unexported variable (but fine obviously within the package).

I'm happy with whatever is clearest to the next guy.

File: services/hh/queue.go (hunk 3 above)
Comment (otoolep) [inline services/hh/queue.go]: OK, that is best, I will use `l.tail`.

File: services/hh/service.go (hunk 4 above)
Comment (otoolep) [inline services/hh/service.go]: I agree, the names here are not great. I was trying to find something that communicated that it's more than just a purge -- it's basically blowing away the queue.

Let me stew on it.

File: services/hh/processor.go (hunk 2 above)
Comment (jwilder) [inline services/hh/processor.go]: I'd say return an error if you call `Remove` and the queue is not already closed.

Comment (otoolep) [inline services/hh/processor.go]: OK, sounds good, I'll do that.

File: services/hh/service.go (hunk 4 above)
Comment (otoolep) [inline services/hh/service.go]: I thnk your suggestion is better, I'll do that.

Comment (jwilder) [inline services/hh/service.go]: I think it's also important to convey that it's only inactive queues somehow.

**Comment by otoolep** [comment]
@jwilder -- there still seems to me to be an issue in `PurgeOlderThan`. If it's called multiple times, a segment will be added multiple times. I see no indication that it is idempotent.

**Comment by otoolep** [comment]
Feedback from @jwilder integrated, merging now.

**Comment by jwilder** [comment]
@otoolep Looks like an existing bug.  Should be straightforward to fix.

**Comment by otoolep** [comment]
OK, I'll take care of it @jwilder.

**Pull request merged**